Initial NEMU CI

sameo opened the pull request:
This is the first basic `semaphoreci` and `pullapprove` based CI for NEMU. We have:

* A `pullapprove` configuration file where 2 LGTMs from the nemu-write team are needed for a PR to be merged.
* A couple of scripts for running minimal CI in a `semaphoreci` context

sameo commented:
@mcastelino @rbradford Please not that we need to get additional privileges in order to enable pullapprove and semaphore CI to gate our wip-master branch.

mcastelino commented on pc-bios/optionrom/code16gcc.h:
```diff
@@ -0,0 +1,3 @@
+asm(
```
@sameo why was this needed?

sameo replied:
The semaphoreci toolchain is gcc 4.8, and  from pc-bios/optionrom/Makefile:

```
ifeq ($(filter -m16, $(QEMU_CFLAGS)),)
# Attempt to work around compilers that lack -m16 (GCC <= 4.8, clang <= ??)
# On GCC we add -fno-toplevel-reorder to keep the order of asm blocks with
# respect to the rest of the code.  clang does not have -fno-toplevel-reorder,
# but it places all asm blocks at the beginning and we're relying on it for
# the option ROM header.  So just force clang not to use the integrated
# assembler, which doesn't support .code16gcc.
QEMU_CFLAGS += $(call cc-option, $(QEMU_CFLAGS), -fno-toplevel-reorder)
QEMU_CFLAGS += $(call cc-option, $(QEMU_CFLAGS), -no-integrated-as)
QEMU_CFLAGS += -m32 -include $(SRC_PATH)/pc-bios/optionrom/code16gcc.h
endif
```

mcastelino commented on tools/CI/minimal_ci.sh:
```diff
@@ -37,6 +37,7 @@ DOWNLOAD_ONLY="false"
 RUN_UNSAFE="false"
 VERBOSE="false"
 CHECK="false"
+VSOCK="true"
```
thanks. I should have done this before. This will also allow folks to test on older host kernels.

sameo commented:
@mcastelino It highly depends on the networking bandwidth, but with a poor one it can take up to 1h.
I will have a look at enabling generic cache for caching the workloads images across builds.
If that does not work, we'll have to strip it down to 2-3 workloads, yes.

sameo commented on tools/CI/semaphore/setup.sh:
```diff
@@ -0,0 +1,11 @@
+
+sudo apt-get install -y libcap-ng-dev librbd-dev
+
+$SRCDIR/tools/build_x86_64.sh
```
> It would also allow the testing of vsock as the ccloudvm scripts use bionic which runs with a kernel new enough to do vsock.

Are you suggesting to also run the actual CI scripts inside ccloudvm?

rbradford replied:
@sameo I run the CI inside ccloudvm myself. We've successfully run qemu inside ccloudvm inside semaphore for CIAO so I think it should work.

rbradford commented on tools/CI/semaphore/setup.sh:
```diff
@@ -0,0 +1,38 @@
+echo ""
+
+go version
+go get github.com/intel/ccloudvm
```
@sameo if you do go get github.com/intel/ccloudvm/... then it will do the right thing with the packages removing the need for the line below.

rbradford commented:
@sameo with the changes i've pushed building and cross building works

I've also also pushed a change to run the CI script inside the ccloudvm, let's see if that works.

sameo commented:
@rbradford Thanks for fixing the ccloudvm based build.
It seems running the CI inside ccloudvm failed.
I cleaned up the branch (I had a debug commit there) and removed your last commit to double check if CI passes now:

https://semaphoreci.com/sameo/nemu/branches/topic-ci/builds/38

rbradford commented:
@sameo I've fixed the PATH inside the script now so it should hopefully find ccloudvm to do the run command.

Irrespective of that it seems to timeout after 1 hour mostly from downloading all the files. Either we could reduce what we do (based off some of our recently discoveries) or maybe use semaphore caching: https://semaphoreci.com/docs/caching-between-builds.html

sameo commented:
@rbradford Thanks a lot, LGTM.
